workflow-manager/limiter/limiter.go

```diff
@@ -0,0 +1,73 @@
+package limiter
+
+import (
+	"fmt"
+	"sync/atomic"
+)
+
+// Limiter is similar to sync.Pool, however, it limtis how many shared
+// objects (in this case workers) there can be
+type Limiter struct {
+	workers    chan int32
+	inProgress int32
+}
+
+// Ticket is the structure that allows you to schedule jobs
+type Ticket struct {
+	worker int32
+	status int32
+}
+
+const (
+	idle    = int32(0)
+	working = int32(1)
+	done    = int32(2)
+)
+
+// New creates a new instance of Limiter that limits how many functions can get executed without the other ones completing first
+func New(maxWorkers int32) *Limiter {
+	// channels with capacity are blocking, setting up a channel with maxWorkers allows us to
+	// use the values inside the channel to coordinate work
+	l := &Limiter{make(chan int32, maxWorkers), 0}
+
+	// These values can be anything, we just use this loop to make them
+	for i := int32(0); i < maxWorkers; i++ {
+		l.workers <- i
+	}
+
+	return l
+}
+
+// RequestTicket requests a ticket for a worker
+// this call will block if there are no tickets available for work
+func (l *Limiter) RequestTicket() *Ticket {
+	worker := <-l.workers
+	return &Ticket{
+		worker: worker,
+		status: int32(0),
+	}
+}
+
+// Execute Executes a function with the work ticket
+func (l *Limiter) Execute(ticket *Ticket, job func()) error {
+	if !atomic.CompareAndSwapInt32(&ticket.status, idle, working) {
+		return fmt.Errorf("did not schedule job - ticket was status")
+	}
+	// It's not important if this gets preempted - this value is for monitoring purposes only
+	atomic.AddInt32(&l.inProgress, 1)
+	job()
+	return nil
+}
+
+// Done signals that the execution for this particular task is complete
+// This call will fail if the Ticket has not been executed
+func (l *Limiter) Done(ticket *Ticket) error {
+	if !atomic.CompareAndSwapInt32(&ticket.status, working, done) {
+		return fmt.Errorf("did not complete job - job was either already completed or ticket was idle")
+	}
```

Comment:
Another advantage of using a `sync.Mutex` is that `Mutex.Lock` cannot fail, which means that `Done()` would no longer need to return error and it would absolve its callers from doing error handling.

Reply:
Yesss! This was really clean.